src/ips.rs

```diff
@@ -0,0 +1,526 @@
+            // located nodes.
+            if self.config.use_geolocation {
+                self.update_rating_by_location(node, &state.nodes, &mut peer_ratings);
+            }
+
+            // Load peerlist with current connections (we don't want to change everything)
+            for (peer_idx, rating) in peer_ratings.iter().enumerate().take(agraph[node_idx].len()) {
+                let peer = agraph[node_idx][peer_idx];
+                peer_list_entry
+                    .list
+                    .push(IpAddr::from_str(state.nodes[peer].ip.as_str()).unwrap());
+
+                // Remeber current peer ratings
+                curr_peer_ratings.push(*rating);
+            }
+
+            // 2 - Calculate desired vertex degree
+            // In the first iteration we will use avg of degree and degree_delta so all
+            // nodes should pursue to degree_delta level. That could be bad if graph's vertexes
+            // have very high (or low) degrees and therefore, delta is very high (or low) too. But until
+            // we have some better idea this one is the best we can do to keep up with the graph.
+            let desired_degree = (degree_avg as u32 + degrees.get(&node_ip).unwrap()) / 2;
+
+            // 3 - Calculate how many peers to add or delete from peerlist
+            let peers_to_delete_count = if desired_degree < *degrees.get(&node_ip).unwrap() {
+                degrees.get(&node_ip).unwrap() - desired_degree
+            } else {
+                // Check if config forces to change peerlist even if we have good degree.
+                // This should be always set to at least one to allow for some changes in graph -
+                // searching for better potential peers.
+                self.config.change_at_least
+            };
+
+            // Calculating how many peers should be added. If we have more peers than desired degree
+            // we will add at least config.change_at_least peers.
+            let mut peers_to_add_count = if desired_degree > *degrees.get(&node_ip).unwrap() {
+                desired_degree - degrees.get(&node_ip).unwrap() + peers_to_delete_count
+            } else {
+                self.config.change_at_least
+            };
+
+            // Limit number of changes to config value
+            if peers_to_add_count > self.config.change_no_more {
+                peers_to_add_count = self.config.change_no_more;
+            }
+
+            // 4 - Choose peers to delete from peerlist (based on ranking)
+            if peers_to_delete_count > 0 {
+                // Sort peers by rating (highest first)
+                curr_peer_ratings.sort_by(|a, b| b.2.partial_cmp(&a.2).unwrap());
+
+                // Remove peers with lowest rating
+                for _ in 0..peers_to_delete_count {
+                    let peer_to_delete = curr_peer_ratings.pop();
+                    if let Some(peer_to_delete) = peer_to_delete {
```

Comment:
if we are deleting, then curr_peer_ratings > peers_to_delete_count so we are safe to use unwrap here directly after `pop()`

Reply:
Now code is changed (that part is refactored) and there is possibility to have curr_peer_ratings empty.

Reply:
Deletion changed one more time to be simple loop with poping values.